index/index_model.go

```diff
@@ -0,0 +1,258 @@
+// allowing or disallowing remote or local file lookups.
+//  - https://github.com/pb33f/libopenapi/issues/73
+type SpecIndexConfig struct {
+    // The BaseURL will be the root from which relative references will be resolved from if they can't be found locally.
+    //
+    // For example:
+    //  - $ref: somefile.yaml#/components/schemas/SomeSchema
+    //
+    // Might not be found locally, if the file was pulled in from a remote server (a good example is the DigitalOcean API).
+    // so by setting a BaseURL, the reference will try to be resolved from the remote server.
+    //
+    // If our baseURL is set to https://pb33f.io/libopenapi then our reference will try to be resolved from:
+    //  - $ref: https://pb33f.io/libopenapi/somefile.yaml#/components/schemas/SomeSchema
+    //
+    // More details on relative references can be found in issue #73: https://github.com/pb33f/libopenapi/issues/73
+    BaseURL *url.URL // set the Base URL for resolving relative references if the spec is exploded.
+
+    // In an earlier version of libopenapi (pre 0.6.0) the index would automatically resolve all references
+    // They could have been local, or they could have been remote. This was a problem because it meant
+    // There was a potential for a remote exploit if a remote reference was malicious. There aren't any known
+    // exploits, but it's better to be safe than sorry.
+    //
+    // To read more about this, you can find a discussion here: https://github.com/pb33f/libopenapi/pull/64
+    AllowRemoteLookup bool // Allow remote lookups for references. Defaults to false
+    AllowFileLookup   bool // Allow file lookups for references. Defaults to false
+
+
+    // private fields
+    seenRemoteSources *syncmap.Map
+    remoteLock        *sync.Mutex
+}
+
+// CreateOpenAPIIndexConfig is a helper function to create a new SpecIndexConfig with the AllowRemoteLookup and
+// AllowFileLookup set to true. This is the default behaviour of the index in previous versions of libopenapi. (pre 0.6.0)
+func CreateOpenAPIIndexConfig() *SpecIndexConfig {
+    return &SpecIndexConfig{
+        AllowRemoteLookup: true,
+        AllowFileLookup:   true,
+        seenRemoteSources: &syncmap.Map{},
+    }
+}
+
+// CreateClosedAPIIndexConfig is a helper function to create a new SpecIndexConfig with the AllowRemoteLookup and
+// AllowFileLookup set to false. This is the default behaviour of the index in versions 0.6.0+
+func CreateClosedAPIIndexConfig() *SpecIndexConfig {
+    return &SpecIndexConfig{
+        AllowRemoteLookup: false,
+        AllowFileLookup:   false,
+        seenRemoteSources: &syncmap.Map{},
+    }
+}
+
+// SpecIndex is a complete pre-computed index of the entire specification. Numbers are pre-calculated and
+// quick direct access to paths, operations, tags are all available. No need to walk the entire node tree in rules,
+// everything is pre-walked if you need it.
+type SpecIndex struct {
```

Comment:
legit; so, this the holy grail of the entire spec.

Reply:
imagine allowing caching this in datastore (s3) in the future

Reply:
It looks like someone can see into the future!